tests/programmability.py

```diff
@@ -86,6 +116,130 @@ def test_custom_endpoints(network, args):
     return network
 
 
+def test_custom_role_definitions(network, args):
+    primary, _ = network.find_primary()
+    member = network.consortium.get_any_active_member()
+
+    # Assign a role to user0
+    user = network.users[0]
+    network.consortium.set_user_data(
+        primary,
+        user.service_id,
+        user_data={"isAdmin": True, "roles": ["ContentGetter"]},
+    )
+
+    content_endpoint_def = {
+        "get": {
+            "js_module": "test.js",
+            "js_function": "content",
+            "forwarding_required": "never",
+            "redirection_strategy": "none",
+            "authn_policies": ["user_cert"],
+            "mode": "readonly",
+            "openapi": {},
+        }
+    }
+
+    bundle_with_auth = {
+        "metadata": {"endpoints": {"/content": content_endpoint_def}},
+        "modules": [{"name": "test.js", "module": TESTJS_ROLE}],
+    }
+
+    # Install app with auth/role support
+    with primary.client(None, None, user.local_id) as c:
+        r = c.put("/app/custom_endpoints", body=bundle_with_auth)
+        assert r.status_code == http.HTTPStatus.NO_CONTENT.value, r.status_code
+
+    # Add role definition
+    prop = member.propose(
+        primary,
+        proposal(
+            action(
+                "set_role_definition", role="ContentGetter", actions=["/content/read"]
```

Comment:
Minor, since this is just a sample, but:

We should really delimit action segments with something other than slashes. This very much looks like a permission for a `/content/read` URI, which doesn't exist. I think I'd be content with a more readable `/content:read` or `read:/content`, and then we could extend our sample to cover multi-element URIs (ie - `read:/content/foo/bar`).

Reply:
@eddyashton I do agree, in fact I'm even suspicious of making this a single string at all, but that's the general convention used in Azure RBAC, as used in TPAL (e.g. `Microsoft.ConfidentialLedger/ledger/governance/constitution/read`). The goal here is to show how this can be done for Azure services, and so staying with that convention is the right thing to do in context.